src/Features/LanguageServer/ProtocolUnitTests/Workspaces/LspWorkspaceManagerTests.cs

```diff
@@ -480,6 +531,164 @@ public async Task TestDoesNotForkWhenDocumentTextBufferOpenedAsync()
+        Assert.Equal("Text", (await document1.GetTextAsync(CancellationToken.None)).ToString());
+
+        // Retrieving the document from lsp should push its content into the workspace.
+        Assert.Equal(
+            (await document1.GetTextAsync(CancellationToken.None)).ToString(),
+            (await testLspServer.TestWorkspace.CurrentSolution.Projects.Single().Documents.Single().GetTextAsync()).ToString());
+
+        // The file should not be in the misc workspace.
+        Assert.Empty(testLspServer.GetManagerAccessor().GetLspMiscellaneousFilesWorkspace()!.CurrentSolution.Projects);
+
+        // Now, if the project system removes the file, we will still see the lsp version of, but back to the misc workspace.
+        await testLspServer.TestWorkspace.ChangeSolutionAsync(
+            testLspServer.TestWorkspace.CurrentSolution.Projects.Single().RemoveDocument(document1.Id).Solution);
+
+        (workspace1, document1) = await GetLspWorkspaceAndDocumentAsync(documentUri, testLspServer).ConfigureAwait(false);
+        Assert.Equal(WorkspaceKind.MiscellaneousFiles, workspace1?.Kind);
+
+        AssertEx.NotNull(document1);
+        Assert.Equal("Text", (await document1.GetTextAsync(CancellationToken.None)).ToString());
+    }
+
+    [Fact]
+    public async Task TestLspDocumentPreferredOverProjectSystemDocumentChangeInMutatingWorkspace()
+    {
+        // This verifies that we will still see the lsp view of the document, even if the project system feeds in
+        // external information about the contents of the file.
+
+        // Start with an empty workspace.
+        await using var testLspServer = await CreateTestLspServerAsync(
+            "Initial Disk Contents", mutatingLspWorkspace: true, new InitializationOptions { ServerKind = WellKnownLspServerKinds.CSharpVisualBasicLspServer });
+
+        var documentUri = testLspServer.TestWorkspace.CurrentSolution.Projects.Single().Documents.Single().GetURI();
+        await testLspServer.OpenDocumentAsync(documentUri, "Text");
+
+        var (workspace, document) = await GetLspWorkspaceAndDocumentAsync(documentUri, testLspServer).ConfigureAwait(false);
+        Assert.Equal(WorkspaceKind.Host, workspace?.Kind);
+
+        // We should see the contents lsp told us about.
+        AssertEx.NotNull(document);
+        Assert.Equal("Text", (await document.GetTextAsync(CancellationToken.None)).ToString());
+
+        // Now, explicitly update the workspace, simulating the project system externally changing it.
+        await testLspServer.TestWorkspace.ChangeSolutionAsync(
+            testLspServer.TestWorkspace.CurrentSolution.WithDocumentText(document.Id, SourceText.From("New Disk Contents")));
+
+        // Querying the workspace directly, prior to an LSP call will show the new contents.
+        document = testLspServer.TestWorkspace.CurrentSolution.Projects.Single().Documents.Single();
+        Assert.Equal("New Disk Contents", (await document.GetTextAsync()).ToString());
```

Comment:
I admit I'm on the fence whether this makes much sense. On one hand, it really ensures we don't get out of sync with LSP. On the other hand, it feels kinda strange to me any host would be letting these sort of changes through in the first place, only to immediately overwrite them again a moment later.